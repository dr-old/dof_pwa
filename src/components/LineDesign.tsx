import { BarChartRounded } from "@mui/icons-material";
import { hexToRgba } from "../utils/helpers";
import { Box, Typography, useTheme } from "@mui/material";
import {
  ChartContainer,
  LinePlot,
  MarkPlot,
  lineElementClasses,
  markElementClasses,
} from "@mui/x-charts";
import { colors } from "../utils/colors";

const pData = [2400, 1398, 9800, 3908, 4800, 3800, 4300];
const xLabels = [
  "Page A",
  "Page B",
  "Page C",
  "Page D",
  "Page E",
  "Page F",
  "Page G",
];

interface ArcDesignProps {
  backgroundColor?: string;
  title?: string;
}
export default function LineDesign({ backgroundColor, title }: ArcDesignProps) {
  const theme = useTheme();

  return (
    <Box
      sx={{
        backgroundColor,
        padding: theme.spacing(2),
        borderRadius: theme.shape.borderRadius,
        textAlign: "center",
        minHeight: 320,
        display: "flex",
        flex: 1,
        flexDirection: "column",
      }}>
      <Box
        sx={{
          display: "flex",
          flexDirection: "row",
          alignItems: "center",
          justifyContent: "space-between",
        }}>
        <Typography
          variant="subtitle1"
          fontWeight={600}
          sx={{ display: "flex", flexDirection: "row" }}>
          <Box
            sx={{
              bgcolor: colors.dark.background.darker,
              width: 30,
              height: 30,
              borderRadius: 2,
              marginRight: 1,
            }}>
            <BarChartRounded fontSize="small" color="warning" />
          </Box>
          {title}
        </Typography>
        <Typography variant="caption">Details</Typography>
      </Box>
      <Box
        sx={{
          my: 2,
          height: 200, // Adjust the height of the chart
          justifyContent: "center",
          alignItems: "center",
          display: "flex",
          flexDirection: "row",
        }}>
        <Box
          sx={{
            display: "flex",
            flexDirection: "column",
            alignItems: "flex-start",
          }}>
          <Typography variant="h4">282%</Typography>
          <Typography
            variant="caption"
            color={hexToRgba(theme.palette.text.primary, 0.5)}>
            Return On Investment
          </Typography>
        </Box>
        <ChartContainer
          width={500}
          height={300}
          series={[{ type: "line", data: pData }]}
          xAxis={[{ scaleType: "point", data: xLabels }]}
          sx={{
            [`& .${lineElementClasses.root}`]: {
              stroke: theme.palette.primary.main,
              strokeWidth: 2,
            },
            [`& .${markElementClasses.root}`]: {
              stroke: theme.palette.primary.main,
              scale: "0.6",
              fill: "#fff",
              strokeWidth: 2,
            },
          }}
          disableAxisListener>
          <LinePlot />
          <MarkPlot />
        </ChartContainer>
      </Box>
    </Box>
  );
}
