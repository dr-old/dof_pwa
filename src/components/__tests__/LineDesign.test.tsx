import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import LineDesign from "../LineDesign";
import { ThemeProvider, createTheme } from "@mui/material/styles";
import { MemoryRouter } from "react-router-dom";

// Mocking the ChartContainer and related components
jest.mock("@mui/x-charts", () => ({
  ChartContainer: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="chart-container">{children}</div>
  ),
  LinePlot: () => <div data-testid="line-plot" />,
  MarkPlot: () => <div data-testid="mark-plot" />,
  lineElementClasses: {
    root: "line-element-root",
  },
  markElementClasses: {
    root: "mark-element-root",
  },
}));

describe("LineDesign Component", () => {
  const theme = createTheme();

  const renderComponent = (backgroundColor?: string, title?: string) =>
    render(
      <ThemeProvider theme={theme}>
        <MemoryRouter>
          <LineDesign backgroundColor={backgroundColor} title={title} />
        </MemoryRouter>
      </ThemeProvider>
    );

  it("renders the title correctly", () => {
    renderComponent("#ffffff", "Test Title");
    expect(screen.getByText("Test Title")).toBeInTheDocument();
  });

  it("renders the chart components", () => {
    renderComponent();
    expect(screen.getByTestId("chart-container")).toBeInTheDocument();
    expect(screen.getByTestId("line-plot")).toBeInTheDocument();
    expect(screen.getByTestId("mark-plot")).toBeInTheDocument();
  });

  it("displays the correct ROI text", () => {
    renderComponent();
    expect(screen.getByText("282%")).toBeInTheDocument();
    expect(screen.getByText("Return On Investment")).toBeInTheDocument();
  });

  it("matches the snapshot", () => {
    const { asFragment } = renderComponent("#ffffff", "Snapshot Title");
    expect(asFragment()).toMatchSnapshot();
  });
});
