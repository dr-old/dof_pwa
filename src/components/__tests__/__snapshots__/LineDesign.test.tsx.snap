// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineDesign Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1ggygw9"
  >
    <div
      class="MuiBox-root css-1ialerq"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle1 css-1sf739h-MuiTypography-root"
      >
        <div
          class="MuiBox-root css-fj66gp"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorWarning MuiSvgIcon-fontSizeSmall css-gyrebt-MuiSvgIcon-root"
            data-testid="BarChartRoundedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 20c1.1 0 2-.9 2-2v-7c0-1.1-.9-2-2-2s-2 .9-2 2v7c0 1.1.9 2 2 2m10-5v3c0 1.1.9 2 2 2s2-.9 2-2v-3c0-1.1-.9-2-2-2s-2 .9-2 2m-4 5c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2s-2 .9-2 2v12c0 1.1.9 2 2 2"
            />
          </svg>
        </div>
        Snapshot Title
      </h6>
      <span
        class="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
      >
        Details
      </span>
    </div>
    <div
      class="MuiBox-root css-p441mq"
    >
      <div
        class="MuiBox-root css-1lxwves"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
        >
          282%
        </h4>
        <span
          class="MuiTypography-root MuiTypography-caption css-1muhxyk-MuiTypography-root"
        >
          Return On Investment
        </span>
      </div>
      <div
        data-testid="chart-container"
      >
        <div
          data-testid="line-plot"
        />
        <div
          data-testid="mark-plot"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
